src/AutoRest.CSharp/Common/Output/Models/Types/HelperTypeProviders/ArgumentProvider.cs

```diff
@@ -0,0 +1,285 @@
+                {
+                    Throw(New.ArgumentOutOfRangeException(_nameParamRef, "Value is less than the minimum allowed.", false))
+                },
+                new IfStatement(LessThan(GetCompareToExpression(new ParameterReference(maxParam), valueParamRef), Literal(0)))
+                {
+                    Throw(New.ArgumentOutOfRangeException(_nameParamRef, "Value is greater than the maximum allowed.", false))
+                }
+            });
+        }
+
+        private ValueExpression GetCompareToExpression(ValueExpression left, ValueExpression right)
+        {
+            return left.Invoke("CompareTo", right);
+        }
+
+        private Method BuildAssertNotDefault()
+        {
+            var valueParam = new Parameter("value", null, _t, null, ValidationType.None, null);
+            var argConstraints = new Dictionary<CSharpType, FormattableString>()
+            {
+                { _t, $"struct, {typeof(IEquatable<>)}" }
+            };
+            var signature = GetSignature("AssertNotDefault", new[] { valueParam.WithRef(), _nameParam }, new[] { _t }, argConstraints);
+            return new Method(signature, new MethodBodyStatement[]
+            {
+                new IfStatement(new BoolExpression(new ParameterReference(valueParam).Invoke("Equals", Default)))
+                {
+                    ThrowArgumentException("Value cannot be empty.")
+                }
+            });
+        }
+
+        private Method BuildAssertNotNullOrWhiteSpace()
+        {
+            var valueParam = new Parameter("value", null, typeof(string), null, ValidationType.None, null);
+            var signature = GetSignature("AssertNotNullOrWhiteSpace", new[] { valueParam, _nameParam });
+            return new Method(signature, new MethodBodyStatement[]
+            {
+                AssertNotNullSnippet(valueParam),
+                new IfStatement(new BoolExpression(new InvokeStaticMethodExpression(typeof(string), "IsNullOrWhiteSpace", new[] { new ParameterReference(valueParam) })))
+                {
+                    ThrowArgumentException("Value cannot be empty or contain only white-space characters.")
+                }
+            });
+        }
+
+        private Method BuildAssertNotNullOrEmptyString()
+        {
+            var valueParam = new Parameter("value", null, typeof(string), null, ValidationType.None, null);
+            var signature = GetSignature(AssertNotNullOrEmptyMethodName, new[] { valueParam, _nameParam });
+            return new Method(signature, new MethodBodyStatement[]
+            {
+                AssertNotNullSnippet(valueParam),
+                new IfStatement(Equal(new MemberExpression(new ParameterReference(valueParam), "Length"), Literal(0)))
+                {
+                    ThrowArgumentException("Value cannot be an empty string.")
+                }
+            });
```

Comment:
I think we should define a `value` expression from the parameter first:
```suggestion
            var signature = GetSignature(AssertNotNullOrEmptyMethodName, new[] { valueParam, _nameParam });
            var value = new StringExpression(valueParam);
            return new Method(signature, new MethodBodyStatement[]
            {
                AssertNotNullSnippet(value),
                new IfStatement(Equal(value.Length, Literal(0)))
                {
                    ThrowArgumentException("Value cannot be an empty string.")
                }
            });
```